primitives/api/proc-macro/src/mock_impl_runtime_apis.rs

```diff
@@ -202,27 +290,42 @@ impl<'a> Fold for FoldRuntimeApiImpl<'a> {
 				&self.impl_trait,
 				&input.sig.ident,
 			);
-			let ret_type = return_type_extract_type(&input.sig.output);
 
-			// Generate the correct return type.
-			input.sig.output = parse_quote!(
-				-> std::result::Result<#crate_::NativeOrEncoded<#ret_type>, Self::Error>
-			);
+			// When using advanced, the user needs to declare the correct return type on its own,
+			// otherwise do it for the user.
+			if !is_advanced {
+				let ret_type = return_type_extract_type(&input.sig.output);
+
+				// Generate the correct return type.
+				input.sig.output = parse_quote!(
+					-> std::result::Result<#crate_::NativeOrEncoded<#ret_type>, Self::Error>
+				);
+			}
 
 			let orig_block = input.block.clone();
 
+			let construct_return_value = if is_advanced {
+				quote!( (move || #orig_block)() )
+			} else {
+				quote! {
+					let __fn_implementation__ = move || #orig_block;
+
+					Ok(#crate_::NativeOrEncoded::Native(__fn_implementation__()))
+				}
+			};
+
 			// Generate the new method implementation that calls into the runtime.
 			parse_quote!(
 				{
 					// Get the error to the user (if we have one).
-					#error
+					#( #errors )*
+
+					let #at_param_name = ___at___param___;
```

Comment:
I guess this name can be shadowed by other param. 
Like:
```rust
		#[advanced]
		fn wild_card(at: &BlockId<Block>, at: u32) -> std::result::Result<NativeOrEncoded<()>, String> {
			Err("Ohh noooo".into())
		}
```
would compile fine but warn that the first at is unused, which I think is fine.

Reply:
I wanted to improve this, but: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dda19a87afbe1b3c862dbe8836ff43a7

It seems that it is valid rust :shrug:

Reply:
(It only prints a warning)